src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -492,17 +450,21 @@ impl<'a, 'tcx> Visitor<'tcx> for Qualifier<'a, 'tcx, 'tcx> {
         match *place {
             Place::Local(ref local) => self.visit_local(local, context, location),
             Place::Static(ref global) => {
-                self.add(Qualif::STATIC);
+                // Only allow statics (not consts) to refer to other statics.
+                if !(self.mode == Mode::Static || self.mode == Mode::StaticMut) {
+                    self.add(Qualif::NOT_CONST);
+                }
```

Comment:
I just noticed this, but if you move this below the `thread_local` check, you can do this:
```rust
                // Only allow statics (not consts) to refer to other statics.
                if self.mode == Mode::Static || self.mode == Mode::StaticMut {
                    return;
                }
                self.add(Qualif::NOT_CONST);
                if self.mode != Mode::Fn {
                    let mut err = struct_span_err!(self.tcx.sess, self.span, E0013,
```
That is, you can rearrange the code such that error reporting is predicated on `self.mode != Mode::Fn`.